File: src/gpu/metal/renderer.m
#import "gpu/metal/renderer.h"
#import "gpu/metal/ShaderTypes.h"
#import "gpu/metal/view.h"
#import "log/manager.h"

@implementation MetalRenderer {
    id<MTLRenderPipelineState> _pipelineState;
    id<MTLCommandQueue> _commandQueue;
    id<MTLBuffer> _vertexBuffer;
    u64 _numVertices;
}

- (instancetype)initWithDevice:(GpuDevice *)device view:(GpuView *)view {
    self = [super init];
    if (self) {
        [self loadMetalWithDevice:device view:view];
    }
    return self;
}

- (void)loadMetalWithDevice:(GpuDevice *)device view:(GpuView *)view {
    LogDefault(gLogManager.logs.gpu, "loadMetal in renderer");
    id<MTLDevice> mtl_device = (__bridge id<MTLDevice>)device;
    ViewDelegate *view_delegate = (__bridge ViewDelegate *)view;
    MTKView *mtk_view = view_delegate.mtk_view;
    id<MTLLibrary> defaultLibrary = [mtl_device newDefaultLibrary];

    id<MTLFunction> vertexFunction = [defaultLibrary newFunctionWithName:@"vertexShader"];
    id<MTLFunction> fragmentFunction = [defaultLibrary newFunctionWithName:@"fragmentShader"];

    MTLRenderPipelineDescriptor *pipelineStateDescriptor = [[MTLRenderPipelineDescriptor alloc] init];
    pipelineStateDescriptor.label = @"Simple Pipeline";
    pipelineStateDescriptor.vertexFunction = vertexFunction;
    pipelineStateDescriptor.fragmentFunction = fragmentFunction;
    pipelineStateDescriptor.colorAttachments[0].pixelFormat = mtk_view.colorPixelFormat;

    NSError *error = NULL;
    _pipelineState = [mtl_device newRenderPipelineStateWithDescriptor:pipelineStateDescriptor
            error:&error];
    if (!_pipelineState) {
        // Pipeline State creation could fail if we haven't properly set up our pipeline descriptor.
        //  If the Metal API validation is enabled, we can find out more information about what
        //  went wrong.  (Metal API validation is enabled by default when a debug build is run
        //  from Xcode)
        LogDefault(gLogManager.logs.gpu,
                "Failed to created pipeline state, error %@", error);
        return;
    }

    NSData *vertexData = [MetalRenderer generateVertexData];

    // Create a vertex buffer by allocating storage that can be read by the GPU
    _vertexBuffer = [mtl_device newBufferWithLength:vertexData.length
                                         options:MTLResourceStorageModeShared];

    // Copy the vertex data into the vertex buffer by accessing a pointer via
    // the buffer's `contents` property
    memcpy(_vertexBuffer.contents, vertexData.bytes, vertexData.length);

    // Calculate the number of vertices by dividing the byte length by the size of each vertex
    _numVertices = vertexData.length / sizeof(Vertex);

    _commandQueue = [mtl_device newCommandQueue];
}

+ (NSData *)generateVertexData
{
    const Vertex quadVertices[] = {
        // Pixel positions, RGBA colors
        {{-20, 20}, {1, 0, 0, 1}},
        {{20, 20}, {0, 0, 1, 1}},
        {{-20, -20}, {0, 1, 0, 1}},

        {{20, -20}, {1, 0, 0, 1}},
        {{-20, -20}, {0, 1, 0, 1}},
        {{20, 20}, {0, 0, 1, 1}},
    };
    const NSUInteger NUM_COLUMNS = 25;
    const NSUInteger NUM_ROWS = 15;
    const NSUInteger NUM_VERTICES_PER_QUAD = sizeof(quadVertices) / sizeof(Vertex);
    const float QUAD_SPACING = 50.0;

    NSUInteger dataSize = sizeof(quadVertices) * NUM_COLUMNS * NUM_ROWS;
    NSMutableData *vertexData = [[NSMutableData alloc] initWithLength:dataSize];

    Vertex* currentQuad = vertexData.mutableBytes;

    for (NSUInteger row = 0; row < NUM_ROWS; row++) {
        for (NSUInteger column = 0; column < NUM_COLUMNS; column++) {
            float2 upperLeftPosition;
            upperLeftPosition.x = ((-((float)NUM_COLUMNS) / 2.0) + column) * QUAD_SPACING + QUAD_SPACING/2.0;
            upperLeftPosition.y = ((-((float)NUM_ROWS) / 2.0) + row) * QUAD_SPACING + QUAD_SPACING/2.0;

            memcpy(currentQuad, &quadVertices, sizeof(quadVertices));

            for (NSUInteger vertexInQuad = 0; vertexInQuad < NUM_VERTICES_PER_QUAD; vertexInQuad++)
            {
                currentQuad[vertexInQuad].position += upperLeftPosition;
            }

            currentQuad += 6;
        }
    }

    return vertexData;
}

- (void)drawInView:(GpuView *)view {
    ViewDelegate *view_delegate = (__bridge ViewDelegate *)view;
    MTKView *mtk_view = view_delegate.mtk_view;
    float2 viewport_size = view_delegate.viewport_size;

    id<MTLCommandBuffer> commandBuffer = [_commandQueue commandBuffer];
    commandBuffer.label = @"MyCommand";

    MTLRenderPassDescriptor *renderPassDescriptor = mtk_view.currentRenderPassDescriptor;

    if (renderPassDescriptor != nil) {
        id<MTLRenderCommandEncoder> renderEncoder =
        [commandBuffer renderCommandEncoderWithDescriptor:renderPassDescriptor];
        renderEncoder.label = @"MyRenderEncoder";

        MTLViewport viewport = {
            0.0,
            0.0,
            viewport_size.x,
            viewport_size.y,
            -1.0,
            1.0,
        };
        [renderEncoder setViewport:viewport];

        [renderEncoder setRenderPipelineState:_pipelineState];

        // We call -[MTLRenderCommandEncoder setVertexBuffer:offset:atIndex:] to send data in our
        //   preloaded MTLBuffer from our ObjC code here to our Metal 'vertexShader' function
        // This call has 3 arguments
        //   1) buffer - The buffer object containing the data we want passed down
        //   2) offset - They byte offset from the beginning of the buffer which indicates what
        //      'vertexPointer' point to.  In this case we pass 0 so data at the very beginning is
        //      passed down.
        //      We'll learn about potential uses of the offset in future samples
        //   3) index - An integer index which corresponds to the index of the buffer attribute
        //      qualifier of the argument in our 'vertexShader' function.  Note, this parameter is
        //      the same as the 'index' parameter in
        //              -[MTLRenderCommandEncoder setVertexBytes:length:atIndex:]
        //
        [renderEncoder setVertexBuffer:_vertexBuffer
                                offset:0
                               atIndex:VertexInputIndexVertices];

        [renderEncoder setVertexBytes:&viewport_size
                               length:sizeof(viewport_size)
                              atIndex:VertexInputIndexViewportSize];

        // Draw the vertices of the quads
        [renderEncoder drawPrimitives:MTLPrimitiveTypeTriangle
                          vertexStart:0
                          vertexCount:_numVertices];

        [renderEncoder endEncoding];

        [commandBuffer presentDrawable:mtk_view.currentDrawable];
    }

    [commandBuffer commit];
}

@end

GpuRenderer *gpu_renderer_create(GpuDevice *device, GpuView *view)
{
    GpuRenderer *renderer;

    @autoreleasepool {
        MetalRenderer *mtl_renderer = [[MetalRenderer alloc]
                initWithDevice:device view:view];
        renderer = (__bridge_retained GpuRenderer *)mtl_renderer;
    }

    return renderer;
}

void gpu_renderer_destroy(GpuRenderer *renderer)
{
    @autoreleasepool {
        MetalRenderer *mtl_renderer = (__bridge_transfer MetalRenderer *)renderer;
        mtl_renderer = nil;
    }
}

void gpu_renderer_draw_in_view(GpuRenderer *renderer, GpuView *view)
{
    @autoreleasepool {
        MetalRenderer *mtl_renderer = (__bridge MetalRenderer *)renderer;
        [mtl_renderer drawInView:view];
    }
}

File: src/gpu/renderer.c
#include "gpu/renderer.h"
#include "gpu/shader-types.h"
#include <stdlib.h>
#include <string.h>

void gpu_renderer_init(GpuRenderer *renderer, GpuDevice *device, GpuView *view, float2 viewport_size)
{
    renderer->viewport_size = viewport_size;
    renderer->vertex_function = gpu_function_create_with_name(device, "vertex_shader");
    renderer->fragment_function = gpu_function_create_with_name(device, "fragment_shader");

    GpuPipelineStateConfig pipeline_state_config = {
        .label = "Simple Pipeline",
        .vertex_function = renderer->vertex_function,
        .fragment_function = renderer->fragment_function,
        .pixel_format = gpu_view_color_pixel_format(view),
    };

    renderer->pipeline_state = gpu_pipeline_state_create(device, &pipeline_state_config);

    // Generate vertex buffer
    {
        const GpuVertex QuadVertices[] = {
            // Pixel positions, RGBA colors
            {{-20, 20}, {1, 0, 0, 1}},
            {{20, 20}, {0, 0, 1, 1}},
            {{-20, -20}, {0, 1, 0, 1}},

            {{20, -20}, {1, 0, 0, 1}},
            {{-20, -20}, {0, 1, 0, 1}},
            {{20, 20}, {0, 0, 1, 1}},
        };
        const i64 NumRows = 15;
        const i64 NumColumns = 25;
        const i64 NumVerticesPerQuad = sizeof(QuadVertices) / sizeof(GpuVertex);
        const float QuadSpacing = 50.0;
        const i64 DataLength = sizeof(QuadVertices) * NumRows * NumColumns;

        renderer->vertex_buffer = gpu_buffer_create_with_length(device,
                DataLength, GpuBufferStorageModeTypeShared);
        renderer->num_vertices = DataLength / sizeof(GpuVertex);

        GpuVertex *current_quad = gpu_buffer_contents(renderer->vertex_buffer);

        for (i32f row = 0; row < NumRows; row++) {
            for (i32f column = 0; column < NumColumns; column++) {
                float2 upper_left_position;
                upper_left_position.x = ((-((float)NumColumns) / 2.0) + column) * QuadSpacing + QuadSpacing/2.0;
                upper_left_position.y = ((-((float)NumRows) / 2.0) + row) * QuadSpacing + QuadSpacing/2.0;

                memcpy(current_quad, &QuadVertices, sizeof(QuadVertices));

                for (i32f vertex_in_quad = 0; vertex_in_quad < NumVerticesPerQuad; vertex_in_quad++)
                {
                    current_quad[vertex_in_quad].position += upper_left_position;
                }

                current_quad += NumVerticesPerQuad;
            }
        }
    }
}

void gpu_renderer_destroy(GpuRenderer *renderer)
{
    gpu_buffer_destroy(renderer->vertex_buffer);
    gpu_pipeline_state_destroy(renderer->pipeline_state);
    gpu_function_destroy(renderer->fragment_function);
    gpu_function_destroy(renderer->vertex_function);
}

void gpu_renderer_draw_in_view(GpuRenderer *renderer, GpuView *view, GpuCommandQueue *queue)
{
    GpuCommandBuffer *command_buffer = gpu_command_buffer_create(queue);
    GpuRenderPassConfig *render_pass_config = gpu_view_current_render_pass_config(view);

    if (render_pass_config) {
        GpuCommandEncoder *encoder = gpu_command_encoder_create(command_buffer, render_pass_config);
        gpu_command_encoder_label(encoder, "MyRenderEncoder");

        gpu_command_encoder_set_viewport(encoder, renderer->viewport_size);

        gpu_command_encoder_set_pipeline_state(encoder, renderer->pipeline_state);

        gpu_command_encoder_set_vertex_buffer(encoder, renderer->vertex_buffer,
                0, GpuVertexIndexVertices);

        gpu_command_encoder_set_vertex_bytes(encoder, &renderer->viewport_size,
                sizeof(renderer->viewport_size), GpuVertexIndexViewportSize);

        gpu_command_encoder_draw_primitives(encoder,
                GpuPrimitiveTypeTriangle, 0, renderer->num_vertices);

        gpu_command_encoder_end_encoding(encoder);
        gpu_command_buffer_present_drawable(command_buffer, view);
    }

    gpu_command_buffer_commit(command_buffer);
}

void gpu_renderer_size_changed(GpuRenderer *renderer, GpuView *view, float2 viewport_size)
{
    (void) view;
    renderer->viewport_size = viewport_size;
}
